### ISSUE TITLE
add middleware/hosts (for reading entries from /etc/hosts file)

### ISSUE THREAD

**Issue by ookangzheng** [coredns/coredns#678]
My settings like this 

proxy . 127.0.0.1:5454 {
        policy round_robin
        fail_timeout 3s
*** lookup custom hosts file (ads block like, 0.0.0.0 ssl.google.com)
        spray
        protocol https_google
}
    cache 160
    log stdout
    errors stdout

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] How can I read my..." ]
>My settings like this
>
>~~~
>proxy . 127.0.0.1:5454 {
>        policy round_robin
>        fail_timeout 3s
>## lookup custom hosts file (ads block like, 0.0.0.0 ssl.google.com)
>        spray
>        protocol https_google
>}
>    cache 160
>    log stdout
>    errors stdout
>~~~

What's the question? You want to be able to block specific sites and getting
that info from the /etc/hosts file? I think we should do that in a separate
middleware, like `hosts` or something that does a lookup in the hosts file
first.

An actual blocking middleware might also be possible, but a `hosts` middleware
seems more useful at the moment, allthough a host file can only deal with a/aaaa
and ptr lookups.


/Miek

--
Miek Gieben

**Comment by ookangzheng** [comment]
Yap, sorry for my english
I have a list of hosts file with 0.0.0.0 xxx.com 
I want it to lookup from hosts first, if match then return 0.0.0.0, else lookup from upstream dns like 8.8.8.8

**Comment by ookangzheng** [comment]
my solution from now I just dnsmasq then forward upstream to coredns

**Comment by miekg** [comment]
I'll open an issue for having a hosts middleware, seems like a nice
addition and maybe someone will have some fun adding it.

On 29 May 2017 12:13 pm, "Oo Kang Zheng" <notifications@github.com> wrote:

my solution from now I just dnsmasq then forward upstream to coredns

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/coredns/coredns/issues/678#issuecomment-304636528>, or mute
the thread

**Comment by miekg** [comment]
Add a new middleware like this is also a good way to get started with coding in CoreDNS

**Comment by miekg** [comment]
Specifically the middleware should:
* parse and reparse /etc/hosts every x seconds or on change
* probably needs `fallthrough` support, just like we did in the kubernetes middleware
* hook up the middleware into CoreDNS

**Comment by ookangzheng** [comment]
example ads block hosts file
https://www.odrive.com/s/fa24307e-b496-4817-9c44-1245d77f8f97-592d3479

**Comment by pmoroney** [comment]
I worked on this a bit last night.
I pulled a lot of inspiration from the auto backend and copied the host file logic from hosts.go in the net package.
I got it to compile and run, but it seems I have the logic wrong for when to return errors and when to use middleware.NextOrFailure. 
I'll read up more about it tonight.

Thank's for suggesting this on twitter, I wanted to get involved and this is a good starting point.
